sw/device/tests/usbdev_swdrive_test.c

```diff
@@ -0,0 +1,148 @@
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+#define USBDEV_BASE_ADDR TOP_EARLGREY_USBDEV_BASE_ADDR
+
+/**
+ * USBDEV handle
+ */
+static dif_usbdev_t usbdev;
+static dif_usbdev_buffer_pool_t buffer_pool;
+
+static const bool verbose = false;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+bool test_main(void) {
+  CHECK(kDeviceType == kDeviceSimVerilator,
+        "This test is not expected to run on platforms other than the "
+        "Verilator simulation. It needs the USB DPI model.");
+
+  LOG_INFO("Running USB SWDRIVE test");
+
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+  pinmux_testutils_init(&pinmux);
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInUsbdevSense,
+      kTopEarlgreyPinmuxInselIoc7));
+
+  CHECK_DIF_OK(
+      dif_usbdev_init(mmio_region_from_addr(USBDEV_BASE_ADDR), &usbdev));
+
+  // Ensure that the USB device has its default bus configuration
+  dif_usbdev_config_t config = {
+      .have_differential_receiver = dif_bool_to_toggle(false),
+      .use_tx_d_se0 = dif_bool_to_toggle(false),
+      .single_bit_eop = kDifToggleDisabled,
+      .pin_flip = dif_bool_to_toggle(false),
+      .clock_sync_signals = kDifToggleEnabled,
+  };
+  CHECK_DIF_OK(dif_usbdev_configure(&usbdev, &buffer_pool, config));
+
+  busy_spin_micros(8u);
+
+  // Desired drive state of output pins
+  //
+  // Note: this is the default 'Idle' state of the bus for a Full Speed device,
+  //       so the DPI model will register our presence but not yet start talking
+  dif_usbdev_phy_pins_drive_t drive = {
+      // Single-ended output
+      .data = true,
+      .se0 = false,
+      // Differential output
+      .dp = true,
+      .dn = false,
+      // Configuration
+      .diff_receiver_enable = false,
+      .dp_pullup_en = true,
+      .dn_pullup_en = false,
+      .output_enable = true,
+  };
+
+  // Waggle the pull up enables through alternating states a few times to
+  // activate the bypass logic in usbdpi.sv; this disconnects the DPI model
+  // within the 'Reset Recovery' time so it doesn't try to communicate with the
+  // device
+  for (unsigned i = 0U; i < 16U; i++) {
+    CHECK_DIF_OK(
+        dif_usbdev_set_phy_pins_state(&usbdev, kDifToggleEnabled, drive));
+
+    drive.dp_pullup_en = !drive.dp_pullup_en;
+    drive.dn_pullup_en = !drive.dn_pullup_en;
+
+    busy_spin_micros(1u);
+  }
+
+  // Delay for a while to separate the 'waggle dance' from the loopback testing;
+  // DPI model enters loopback mode
+  busy_spin_micros(4u);
+
+  // Sensed state of input pins
+  dif_usbdev_phy_pins_sense_t sense;
+
+  // Establish a loopback path; requires DPI assistance because usbdev
+  // has physically separate input and output buses
+  for (unsigned d = 0u; d < 0x100u; d++) {
+    drive.dp = ((d & 1u) != 0u);
```

Comment:
I think the `bitfield_bit32` API should be used with the auto generated macros to improve readability:
https://github.com/lowRISC/opentitan/blob/65783deea15fa214a067742848f6092b832eaeaf/sw/device/lib/dif/dif_usbdev.c#L899-L909

Reply:
We discussed that offline, a long time ago now. This test is treating the phy outputs and inputs as regular GPIO pins, but it must use the DIF interface to access the hardware, so it's making from bits to structure fields (which map to signal/pin names), and the DIF is what contains the above suggested code for mapping from fields to register bits.